clouddriver-appengine/src/main/groovy/com/netflix/spinnaker/clouddriver/appengine/provider/agent/AppEngineServerGroupCachingAgent.groovy

```diff
@@ -0,0 +1,438 @@
+
+import com.fasterxml.jackson.core.type.TypeReference
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.google.api.client.googleapis.batch.BatchRequest
+import com.google.api.client.googleapis.json.GoogleJsonError
+import com.google.api.client.http.HttpHeaders
+import com.google.api.services.appengine.v1.model.Instance
+import com.google.api.services.appengine.v1.model.ListInstancesResponse
+import com.google.api.services.appengine.v1.model.ListVersionsResponse
+import com.google.api.services.appengine.v1.model.Service
+import com.google.api.services.appengine.v1.model.Version
+import com.netflix.frigga.Names
+import com.netflix.spectator.api.Registry
+import com.netflix.spinnaker.cats.agent.AgentDataType
+import com.netflix.spinnaker.cats.agent.CacheResult
+import com.netflix.spinnaker.cats.agent.DefaultCacheResult
+import com.netflix.spinnaker.cats.cache.CacheData
+import com.netflix.spinnaker.cats.cache.DefaultCacheData
+import com.netflix.spinnaker.cats.provider.ProviderCache
+import com.netflix.spinnaker.clouddriver.appengine.AppEngineCloudProvider
+import com.netflix.spinnaker.clouddriver.appengine.cache.Keys
+import com.netflix.spinnaker.clouddriver.appengine.model.AppEngineInstance
+import com.netflix.spinnaker.clouddriver.appengine.model.AppEngineLoadBalancer
+import com.netflix.spinnaker.clouddriver.appengine.model.AppEngineServerGroup
+import com.netflix.spinnaker.clouddriver.appengine.provider.callbacks.AppEngineCallback
+import com.netflix.spinnaker.clouddriver.appengine.security.AppEngineNamedAccountCredentials
+import com.netflix.spinnaker.clouddriver.cache.OnDemandAgent
+import com.netflix.spinnaker.clouddriver.cache.OnDemandMetricsSupport
+import com.netflix.spinnaker.clouddriver.kubernetes.provider.view.MutableCacheData
+import groovy.util.logging.Slf4j
+
+import static com.netflix.spinnaker.cats.agent.AgentDataType.Authority.AUTHORITATIVE
+import static com.netflix.spinnaker.clouddriver.appengine.cache.Keys.Namespace.APPLICATIONS
+import static com.netflix.spinnaker.clouddriver.appengine.cache.Keys.Namespace.CLUSTERS
+import static com.netflix.spinnaker.clouddriver.appengine.cache.Keys.Namespace.INSTANCES
+import static com.netflix.spinnaker.clouddriver.appengine.cache.Keys.Namespace.LOAD_BALANCERS
+import static com.netflix.spinnaker.clouddriver.appengine.cache.Keys.Namespace.SERVER_GROUPS
+import static com.netflix.spinnaker.clouddriver.appengine.cache.Keys.Namespace.ON_DEMAND
+
+@Slf4j
+class AppEngineServerGroupCachingAgent extends AbstractAppEngineCachingAgent implements OnDemandAgent {
+  final String category = "serverGroup"
+
+  final OnDemandMetricsSupport metricsSupport
+
+  static final Set<AgentDataType> types = Collections.unmodifiableSet([
+    AUTHORITATIVE.forType(APPLICATIONS.ns),
+    AUTHORITATIVE.forType(CLUSTERS.ns),
+    AUTHORITATIVE.forType(SERVER_GROUPS.ns),
+    AUTHORITATIVE.forType(INSTANCES.ns),
+    AUTHORITATIVE.forType(LOAD_BALANCERS.ns),
```

Comment:
We talked about this and, because of the way the API is structured, decided that we can't think of a good way of splitting out the API calls into separate caching agents.